src/commonMain/kotlin/org/kosat/CDCL.kt

```diff
@@ -1,24 +1,62 @@
 package org.kosat
 
+import org.kosat.heuristics.Preprocessor
+import org.kosat.heuristics.Restarter
+import org.kosat.heuristics.VariableSelector
+import org.kosat.heuristics.VSIDS
 import kotlin.math.abs
-import kotlin.math.max
 
 // CDCL
 fun solveCnf(cnf: CnfRequest): List<Int>? {
-    val clauses = (cnf.clauses.map { it.lit }).toMutableList()
-    return CDCL(clauses, cnf.vars).solve()
+    val clauses = (cnf.clauses.map { it.lits }).toMutableList()
+    return CDCL(clauses.map { Clause(it) }.toMutableList(), cnf.vars).solve()
 }
 
+enum class SolverType {
+    INCREMENTAL, NON_INCREMENTAL;
+}
 
-class CDCL(val clauses: MutableList<MutableList<Int>>, initNumber: Int = 0) {
-    var varsNumber = initNumber
-        private set
+class CDCL(private val solverType: SolverType = SolverType.INCREMENTAL) : Incremental {
 
-    init {
-        // set varsNumber equal to either initNumber(from constructor of class) either maximal variable from cnf
-        varsNumber =
-            max(initNumber, clauses.flatten().let { all -> if (all.isNotEmpty()) all.maxOf { abs(it) } else 0 })
-    }
+    // we never store clauses of size 1
+    // they are lying at 0 decision level of trail
+
+    // initial constraints + externally added by newClause
+    val constraints = mutableListOf<Clause>()
+    // learnt from conflicts clauses, once in a while their number halved
+    val learnts = mutableListOf<Clause>()
+    var varsNumber: Int = 0
+
+    // contains current assignment, clause it came from and decision level when it happened
+    val vars: MutableList<VarState> = MutableList(varsNumber + 1) { VarState(VarStatus.UNDEFINED, null, -1) }
+
+    // all decisions and consequences, contains variables
+    private val trail: MutableList<Int> = mutableListOf()
+
+    // two watched literals heuristic; in watchers[i] set of clauses watched by variable i
+    private val watchers = MutableList(varsNumber * 2 + 1) { mutableListOf<Clause>() }
```

Comment:
Consider `Watcher` class.

I thought this PR is trying to merge the `classes` branch, which was all about the "class-ification" of the codebase. So why not wrap the watcher in a separate class?

```kotlin
class Watcher(
    val clause: Clause,
    val blocker: Lit,
)
```

btw, do NOT move the `updateWatcher` inside a `Watcher` class (as you might thought) — it is better to keep the `updateWatchers` as a Solver's method.